nibabel/tests/test_funcs.py

```diff
@@ -30,34 +30,96 @@ def _as_fname(img):
-        img_files = [_as_fname(img) for img in imgs]
-        for img0, img1, img2, img3 in (imgs, img_files):
-            all_imgs = concat_images([img0, img1])
-            assert_array_equal(all_imgs.get_data(), all_data)
-            assert_array_equal(all_imgs.affine, affine)
-            # check that not-matching affines raise error
-            assert_raises(ValueError, concat_images, [img0, img2])
-            assert_raises(ValueError, concat_images, [img0, img3])
-            # except if check_affines is False
-            all_imgs = concat_images([img0, img1])
-            assert_array_equal(all_imgs.get_data(), all_data)
-            assert_array_equal(all_imgs.affine, affine)
-        # Delete images as prophylaxis for windows access errors
-        for img in imgs:
-            del(img)
+    for dim in range(2, 6):
+        all_shapes_ND = tuple((shape[:dim] for shape in all_shapes_5D))
+        all_shapes_N1D_unary = tuple((shape + (1,) for shape in all_shapes_ND))
+        all_shapes = all_shapes_ND + all_shapes_N1D_unary
+
+        # Loop over all possible combinations of images, in first and
+        #   second position.
+        for data0_shape in all_shapes:
+            data0_numel = np.asarray(data0_shape).prod()
+            data0 = np.arange(data0_numel).reshape(data0_shape)
+            img0_mem = Nifti1Image(data0, affine)
+
+            for data1_shape in all_shapes:
+                data1_numel = np.asarray(data1_shape).prod()
+                data1 = np.arange(data1_numel).reshape(data1_shape)
+                img1_mem = Nifti1Image(data1, affine)
+                img2_mem = Nifti1Image(data1, affine+1)  # bad affine
+
+                # Loop over every possible axis, including None (explicit and implied)
+                for axis in (list(range(-(dim-2), (dim-1))) + [None, '__default__']):
+
+                    # Allow testing default vs. passing explicit param
+                    if axis == '__default__':
+                        np_concat_kwargs = dict(axis=-1)
+                        concat_imgs_kwargs = dict()
+                        axis = None  # Convert downstream
+                    elif axis is None:
+                        np_concat_kwargs = dict(axis=-1)
+                        concat_imgs_kwargs = dict(axis=axis)
+                    else:
+                        np_concat_kwargs = dict(axis=axis)
+                        concat_imgs_kwargs = dict(axis=axis)
+
+                    # Create expected output
+                    try:
+                        # Error will be thrown if the np.concatenate fails.
+                        #   However, when axis=None, the concatenate is possible
+                        #   but our efficient logic (where all images are
+                        #   3D and the same size) fails, so we also
+                        #   have to expect errors for those.
+                        expect_error = data0.ndim != data1.ndim
```

Comment:
I personally try to keep indent levels to be a logical block, so sometimes I include things inside a `try...except` that may not throw, so that the logical block is maintained.  Here, the computation of `expect_error` is all at the same indent level, which I find to be semantically helpful.

With that said, I'm happy to move it out; doing that now!